FormidableLabs/spectacle
# MDX Slide Parsing and CLI

### Description

This PR introduces a two new pieces of functionality:
* MDX file parsing that delineates Spectacle `<Slide />`s by the presence of `---` in the MDX
*  CLI functionality that can instantly boot up a `webpack-dev-server` instance for an arbitrary MDX file

These two features work hand in hand to simplify how MDX files can be used in Spectacle. Like in Old Spectacle, `---` can be used to denote separate slides. This will work separately from the CLI. However, in combination with the CLI, booting up a hot-reloading Spectacle presentation from an MDX file is very straightforward. Once Spectacle is installed globally (or `npm link`'d from the cloned repo), a user can run this from any arbitrary directory:
`spectacle -s <name of file>`.

The `-s` flag tells the CLI that the user wants to boot up a presentation from the provided file. The CLI defaults to looking for a file named `slides.mdx`, but the user can give an alternative name/path. For example:
`spectacle` (looks for `slides.mdx` in the current directory) or `spectacle -s ./nested/my-pres.mdx` (points Spectacle directly to the MDX file). In the future, other file types might be supported with the `-s` option. For now, only MDX files are supported.

**Webpack** - when you see the webpack configuration, you might wonder why it requires so many changes. It appears that when webpack is run from a directory outside of the Spectacle project (i.e. as a globally installed CLI tool), it has immense trouble finding the loaders and babel tools that are used in the configuration. The only solution that I found to bypass this problem was to use Node's `path` library to construct paths to the dependencies.

**MDX Loader** - the loader file has comments to explain each step. It takes significant inspiration from the custom MDX loader in Old Spectacle. However, it includes updates that are a necessity born of the `mdx-js` version that we're using.


![ezgif com-video-to-gif-2](https://user-images.githubusercontent.com/6595948/62887135-e8324d00-bcf9-11e9-977a-55d455e36f07.gif)

#### Type of Change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

### How Has This Been Tested?

To test new functionality:
1. check out this branch locally
2. run `npm link` in the spectacle directory
3. move to any directory you like, e.g. `~/Documents/test/nest`
4. create a new MDX file with valid content (see https://github.com/FormidableLabs/spectacle/pull/704/files#diff-5e02b1a431b9e43c86f7a48c7a90294a as an example); remember that you must create any components that are imported into the MDX
5. run `spectacle -s <your mdx file>` to boot up the Spectacle presentation
